### PR TITLE
avoid redundant buffer allocations in kernel generator

### PR DESCRIPTION
**Pull request by t4c1**
## Summary
In `operation_cl` the `matrix_cl` conversion operator always constructs new `matrix_cl` for result. For simple kernel with result matrix already of correct size this is considerable overhead.

I replaced this conversion operator with a consturctor and assignment operator on `matrix_cl` to avoid these redundant allocations.

## Tests
No tests as this is just a refactor.

## Side Effects
For some reason I don't understand this change broke `matrix_cl::wait_for_read_write_events()`. I guess `EnquequeBarrierWithWaitList` might be buggy. Anyway my reimplementation that avoids calling this function is simpler so I think no harm is done here.

## Checklist

- [x] Math issue #1342 

- [x] Copyright holder: Tadej Ciglarič
    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/opencl/kernel_generator/is_valid_expression.hpp [hunk 1]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_OPENCL_IS_USABLE_AS_OPERATION
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/is_valid_expression.hpp]: Not added in this PR but the header of this file doesnt match the filename.

File: stan/math/opencl/kernel_generator/matrix_cl_conversion.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_MATRIX_CL_CONVERSION_H
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_MATRIX_CL_CONVERSION_H
```
Comment (rok-cesnovar) [inline stan/math/opencl/kernel_generator/matrix_cl_conversion.h]: Why is this file .h?

Comment (t4c1) [reply stan/math/opencl/kernel_generator/matrix_cl_conversion.h]: Due to some editior default i did not notice. Fixed.

File: stan/math/opencl/kernel_generator/is_valid_expression.hpp (hunk 1 above)
Comment (t4c1) [reply stan/math/opencl/kernel_generator/is_valid_expression.hpp]: Fixed this and some other header guards.

**Pull request merged**